keps/sig-storage/177-volume-snapshot-validation/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+
+A webhook server receives AdmissionReview(definition) requests from API server, and responses with a response of the same type to either admit/deny the request. Following simplified diagram shows the workflow.
+(Note that the mutatting webhooks will be invoked BEFORE validating).
+
+The webhook server will expose an HTTP endpoint such that to allow the API server to send AdmissionReview requests. Webhook server providers can dynamically(details) configure what type of resources and what type of admission webhooks via creating CRs of type ValidatingWebhookConfiguration and/or MutatingWebhookConfiguration. It’s worth thinking to provide a generic webhook server in csi-repo. At this stage, the proposal is to create a repo under CSI org, however only implement the validating webhooks for VolumeSnapshot.
+
+CRD validation is preferred over webhook validation due to their lower complexity, however CRD validation schema is unable to enforce immutability or provide ratcheting validation.
+
+### Goals
+
+- Provide an updated CRD schema to validate fields
+- Prevent:
+  - Invalid VolumeSnapshot/VolumeSnapshotContent from creation and update
+  - Invalid updates on immutable fields, i.e., VolumeSnapshot.Spec.Source
+- Provide a pre-built image which can be used to deploy the server
+- Provide a way to deploy the webhook server in cluster
+- Provide a way to authenticate the webhook server to the API server via TLS
+- Provide a release process to safely tighten the validation and move towards the ideal state of using builtin CRD validation while maintaining backwards compatibility
+
+### Non-Goals
+
+- Provide a way to authenticate the API server to the webhook server
+
+## Proposal
+
+Tighten the validation on Volume Snapshot objects. The following fields will begin to enforce immutability: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`. The following fields will begin to enforce `oneOf`: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`). The following fields will begin to enforce non-empty strings: `spec.VolumeSnapshotClassName`. More details are in the Validating Scenarios section.
+
+Due to backwards compatibility concerns, the tightening will occur in two phases.
+
+1. The first phase is webhook-only, and will use ratcheting validation combined with a reconciliation method to delete or fix currently persisted objects which are invalid under the new (strict) validation rules.
+2. The second phase occurs once all invalid objects are cleared from the cluster. The CRD schema validation will be tightened and the webhook will stick around to enforce immutability until immutable fields come to CRDs. (or the crd upgrade could wait until immutable fields are available to do in one go)
+
+The phases come in separate releases to allow users / cluster admin the opportunity to clean their cluster of any invalid objects. More details are in the Risks and Mitigations section.
```

Comment:
Is 1 release enough? I'm not quite sure where this falls under the kubernetes deprecation policy. This is changing the behavior of a beta API, but the deprecation policy mostly talks about removing entire API groups. cc @liggitt

Reply:
Most of this tightening is like fixing bugs. The oneof restriction is already handled by the controller.  So I think we should not wait for more than 1 release.